mlflow/torchscript.py

```diff
@@ -0,0 +1,125 @@
+        # If it walks like a duck and it quacks like a duck, then it must be a duck
+        raise TypeError("Argument 'model' should be a TorchScript model")
+    path = Path(path)
+    try:
+        path.mkdir(parents=True, exist_ok=False)
+    except FileExistsError:
+        raise RuntimeError("Path '{}' already exists".format(path))
+
+    if mlflow_model is None:
+        mlflow_model = Model()
+
+    if signature is not None:
+        mlflow_model.signature = signature
+
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, path)
+
+    filepath = str(path / _SERIALIZED_TORCH_MODEL_FILE_NAME)
+    model.save(filepath, **kwargs)
+    conda_env_subpath = "conda.yaml"
+    if conda_env is None:
+        conda_env = get_default_conda_env()
+    elif not isinstance(conda_env, dict):
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+    with open(path / conda_env_subpath, "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+    mlflow_model.add_flavor(FLAVOR_NAME, pytorch_version=torch.__version__)
+    pyfunc.add_to_model(mlflow_model, loader_module="mlflow.torchscript", env=conda_env_subpath)
+    mlflow_model.save(path / "MLmodel")
+
+
+def load_model(model_uri, **kwargs):
+    local_model_path = Path(_download_artifact_from_uri(artifact_uri=model_uri))
+    torchscript_conf = _get_flavor_configuration(model_path=local_model_path,
+                                             flavor_name=FLAVOR_NAME)
+    if torch.__version__ != torchscript_conf["pytorch_version"]:
+        _logger.warning(
+            "Stored model version '%s' does not match installed PyTorch version '%s'",
+            torchscript_conf["pytorch_version"], torch.__version__)
+    return torch.jit.load(
+        str(local_model_path / _SERIALIZED_TORCH_MODEL_FILE_NAME), **kwargs)
+
+
+def _load_pyfunc(path, **kwargs):
+    """
+    Load PyFunc implementation. Called by ``pyfunc.load_pyfunc``.
+
+    :param path: Local filesystem path to the MLflow Model with the ``torchscript`` flavor.
+    """
+    loaded = load_model(path, **kwargs)
+    return _PyTorchWrapper(loaded)
+
+
+class _PyTorchWrapper(object):
+    """
+    Wrapper class that creates a predict function such that
+    predict(data: pd.DataFrame) -> model's output as pd.DataFrame (pandas DataFrame)
+    """
+    def __init__(self, torchscript_model):
+        self.torchscript_model = torchscript_model
+
+    def predict(self, data, device='cpu'):
+        if not isinstance(data, pd.DataFrame):
+            raise TypeError("Input data should be pandas.DataFrame")
+        self.torchscript_model.to(device)
+        self.torchscript_model.eval()
+        with torch.no_grad():
+            input_tensor = torch.from_numpy(data.values.astype(np.float32)).to(device)
+            preds = self.torchscript_model(input_tensor)
+            if not isinstance(preds, torch.Tensor):
```

Comment:
BTW, is there any way to validate that the Pytorch model produces a single output tensor in `save_model` and raise an exception there? It seems like it could be a bit frustrating for the user to save the model only to find that at inference time that it cannot be scored as a pyfunc. Adding such validation in `save_model` has the downside that you can't e.g. run code like:

```
# Save model
mlflow.torchscript.save_model(model_with_2d_output, path)
# Load model back as a torch.jit model rather than as a pyfunc
loaded_model = mlflow.torchscript.load_model(path)
```

for users who aren't interested in the pyfunc flavor. But I think it'd be better to be consistent here / fail at saving time if we can't use the model with the pyfunc flavor, given that much of MLflow's built-in serving tools rely on the pyfunc flavor

Reply:
IMO, that's a hard (and avoidable) restriction to put on users.
In any case, I don't know a way to check this reliably. What we could probably do is to mention this in the docstring explicitly as `mlflow.pytorch` does